File: src/pvz_console/parsers/almanac.py
from __future__ import annotations

from typing import List

from pvz_console.core.models import Achievement, Plant, Zombie
from pvz_console.parsers.loaders import load_json, source_almanac_path

PLANTS_FILE = "LawnStringsTranslate.json"
ZOMBIES_FILE = "ZombieStringsTranslate.json"
ACHIEVEMENTS_FILE = "AchievementsTextTranslate.json"


def load_plants(root: str, locale: str) -> List[Plant]:
    data = load_json(source_almanac_path(root, locale, PLANTS_FILE))
    return [
        Plant(id=item.get("seedType"), name=item.get("name"), raw=item)
        for item in data.get("plants", [])
        if isinstance(item, dict)
    ]


def load_zombies(root: str, locale: str) -> List[Zombie]:
    data = load_json(source_almanac_path(root, locale, ZOMBIES_FILE))
    return [
        Zombie(id=item.get("theZombieType"), name=item.get("name"), raw=item)
        for item in data.get("zombies", [])
        if isinstance(item, dict)
    ]


def load_achievements(root: str, locale: str) -> List[Achievement]:
    data = load_json(source_almanac_path(root, locale, ACHIEVEMENTS_FILE))
    return [
        Achievement(id=item.get("achievement"), name=item.get("Name"), raw=item)
        for item in data.get("achievements", [])
        if isinstance(item, dict)
    ]

File: src/parsers/almanac.ts
import type { Achievement, Id, Plant, Zombie } from "../core/models";
import { loadJson, sourceAlmanacPath } from "./loaders";

export const PLANTS_FILE = "LawnStringsTranslate.json";
export const ZOMBIES_FILE = "ZombieStringsTranslate.json";
export const ACHIEVEMENTS_FILE = "AchievementsTextTranslate.json";

type RawEntry = Record<string, unknown>;

function isRecord(value: unknown): value is RawEntry {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/** Mirrors Python's `dict.get` (absent → `None`). */
function get(item: RawEntry, key: string): unknown {
  return key in item ? item[key] : null;
}

function collection(data: unknown, key: string): RawEntry[] {
  if (!isRecord(data)) return [];
  const items = data[key];
  if (!Array.isArray(items)) return [];
  return items.filter(isRecord);
}

export function loadPlants(root: string, locale: string): Plant[] {
  const data = loadJson(sourceAlmanacPath(root, locale, PLANTS_FILE));
  return collection(data, "plants").map((item) => ({
    id: get(item, "seedType") as Id,
    name: get(item, "name") as string | null,
    raw: item,
  }));
}

export function loadZombies(root: string, locale: string): Zombie[] {
  const data = loadJson(sourceAlmanacPath(root, locale, ZOMBIES_FILE));
  return collection(data, "zombies").map((item) => ({
    id: get(item, "theZombieType") as Id,
    name: get(item, "name") as string | null,
    raw: item,
  }));
}

export function loadAchievements(root: string, locale: string): Achievement[] {
  const data = loadJson(sourceAlmanacPath(root, locale, ACHIEVEMENTS_FILE));
  return collection(data, "achievements").map((item) => ({
    id: get(item, "achievement") as Id,
    name: get(item, "Name") as string | null,
    raw: item,
  }));
}
